thermal-control-design.md

```diff
@@ -0,0 +1,272 @@
+ |:---:|:-----------:|:------------------:|-----------------------------------|
+ | 0.1 |             |      Liu Kebo      | Initial version                   |
+ | 0.2 |             |      Liu Kebo      | Revised after community review    |
+
+
+  
+## 1. Overview
+
+The purpose of Thermal Control is to keep the switch at a proper temperature by using cooling devices, e.g., fan.
+Thermal control daemon need to monitor the temperature of devices (CPU, ASIC, optical modules, etc) and the running status of fan. It store temperature values fetched from sensors and thermal device running status to the DB, to make these data available to CLI and SNMP or other apps which interested. 
+
+Thermal control also enforce some environment related polices to help the thermal control algorithm to adjust the switch temperature.     
+
+## 2. Thermal device monitoring
+
+Thermal monitoring function will retrieve the switch device temperatures via platform APIs, platform APIs not only provide the value of the temperature, but also provide the threshold value. The thermal object status can be deduced by comparing the current temperature value against the threshold. If it above the high threshold or under the low threshold, alarm shall be raised.
+
+Besides device temperature, shall also monitoring the fan running status.
+
+Thermal device monitoring will loop at a certain period, 60s can be a good value since usually temperature don't change much in a short period.
+
+### 2.1 Temperature monitoring 
+
+In new platform API ThermalBase() class provides get_temperature(), get_high_threshold(),  get_low_threshold(), get_critical_high_threshold() and get_critical_low_threshold() functions, values for a thermal object can be fetched from them. Warning status can also be deduced. 
+
+For the purpose of feeding CLI/SNMP or telemetry functions, these values and warning status can be stored in the state.  DB schema can be like this:
+
+    ; Defines information for a thermal object
+    key                     = TEMPERATURE_INFO|object_name   ; name of the thermal object(CPU, ASIC, optical modules...)
+    ; field                 = value
+    temperature             = FLOAT                          ; current temperature value
+    timestamp               = STRING                         ; timestamp for the temperature fetched
+    high_threshold          = FLOAT                          ; temperature high threshold
+    critical_high_threshold = FLOAT                          ; temperature critical high threshold
+    low_threshold           = FLOAT                          ; temperature low threshold
+    critical_low_threshold  = FLOAT                          ; temperature critical low threshold
+    warning_status          = BOOLEAN                        ; temperature warning status
+
+These devices shall be included to the temperature monitor list but not limited to: CPU core, CPU pack, ASIC, PSU, Optical Modules, etc.
```

Comment:
Can you confirm if ASIC monitoring here refers to the external sensors placed in proximity to ASIC or ASIC internal diodes ? I was planning to add a poller (running in orchagent:SwitchOrch context) which will collect SAI_SWITCH_ATTR_MAX_TEMP, SAI_SWITCH_ATTR_TEMP_LIST, SAI_SWITCH_ATTR_AVERAGE_TEMP and dump to this TEMPERATURE_INFO table (since the other external thermal sensor values may be retrieved as part of standard Platform API calls instead of via SAI)...

Reply:
thermal value is reading from the thermal API, in this layer we don't know exactly where the sensor is located, it could be various per vendors, let me know how these SAI temp defined(in terms which sensor it polling), I can check internally.

Reply:
Understood. The sensors I was referring to are ASIC internal sensors. I get that currently the Platform APIs don't make SAI calls. So, it has to be a separate poller.